spatialos-sdk/src/worker/parameters.rs

```diff
@@ -430,3 +408,112 @@ impl ThreadAffinityParameters {
         }
     }
 }
+
+/// Helper struct for converting `ConnectionParameters` into `Worker_ConnectionParameters`.
+pub(crate) struct IntermediateConnectionParameters<'a> {
+    params: &'a ConnectionParameters,
+    protocol: IntermediateProtocolType,
+}
+
+impl<'a> IntermediateConnectionParameters<'a> {
+    pub(crate) fn as_raw(&self) -> Worker_ConnectionParameters {
```

Comment:
There's no guarantee that the `Worker_ConnectionParameters` lives as long as `&self`, so you can end up with dangling references here, right? 

Wouldn't it be more correct to construct the `Worker_ConnectionParameters` when you instantiate `IntermediateConnectionParameters` and only allow it to be borrowed? Thus tying the lifetimes of `ConnectionParameters` and the C representation? 

I.e. - 

```rust
pub(crate) struct IntermediateConnectionParameters<'a> {
    params: &'a ConnectionParameters,
    protocol: IntermediateProtocolType,
	raw_params: Worker_ConnectionParameters
}

impl<'a> IntermediateConnectionParameters<'a> {
	pub fn new(params: &'a ConnectionParameters, protocol: IntermediateProtocolType) -> Self{
		IntermediateConnectionParameters {
			params,
			protocol,
			Self::to_raw(params)
		}
	}

	pub fn raw(&self) -> &Worker_ConnectionParameters { &self.raw_params }

	fn to_raw(params: &ConnectionParameters) -> Worker_ConnectionParameters { ... }
}

Reply:
Problem is that `Worker_ConnectionParameters` is `Clone`/`Copy`, so it would still be trivial to create a copy of the `Worker_ConnectionParameters` that ends up outliving the `IntermediateConnectionParameters` that owns the data it points to.

As far as I can tell, there's no way to statically prevent misuse of `Worker_ConnectionParameters` (or any other C type with pointers in it). Since it uses raw pointers instead of references, it doesn't adhere to Rust's ownership systems and so the lifetimes are effectively invisible to the compiler. The best we can do is make something that is:

1. Easier to use correctly than not.
2. Easy to visually verify is correct when reviewing changes.

The one nice thing about this approach (as opposed to the closure-based approach) is that you never have to actually bind the `Worker_ConnectionParameters` to a variable. You can always pass `&intermediate.as_raw()` to the C API, which helps reduce the likelihood that you try to keep the `Worker_ConnecitonParameters` around longer than is valid. However, this is still only a convention and can't be enforced by the compiler 😞

Reply:
I wonder if we can special case `Worker_ConnectionParameters` to not derive `Clone`/`Copy` on it. If so, we should be able to eliminate the risk (however low it may be)

Reply:
There's many occurrences of this sort of thing throughout the codebase, it would be nice to be able to find a pattern that we can use to get as close to guaranteed safety as possible (up to stopping auto-derived traits for types, or even a blanket ban).

`Copy` in particular on a structure with pointers feels dangerous 😬

Reply:
YOLO'ed together a trait-struct combo based on the above that we could try to use as a pattern if we had no `Copy/Clone` on the generated bindings. (No guarantee that it compiles 😉)

```rust
// Assuming we have no Copy/Clone on "U" the following struct and trait should provide a safe binding 
// between the Rust (safe) representation and the C (unsafe) representation.

pub(crate) trait WorkerSdkConversion<U> {
    fn raw(&Self) -> U;
}

pub(crate) struct RawBinding<'a, T, U> {
    safe: &'a T,
    raw: U
}

impl<'a, T, U> RawBinding<'a, T, U> 
    where T : WorkerSdkConversion<U>,
    /* U: !Copy + !Clone                if we had negative type constraints */
    {
        pub fn new(safe: &'a T) -> Self {
            Self {
                safe,
                raw: safe.raw()
            }
        }

        pub fn as_raw(&self) -> &U {
            &self.raw
        }
}

impl WorkerSdkConversion<Worker_ConnectionParameters> for ConnectionParameters {
    pub fn raw(&self) -> Worker_ConnectionParameters {
        ...
    }
}

impl WorkerConnection {
    pub fn make_connection(params: &ConnectionParameters) -> Future<WorkerConnection> {
        let binding = RawBinding::new(&params);
        let raw = binding.as_raw();

        drop(binding);
        connect(..., raw); // Invalid!
    }
}
```

Reply:
> Of course, nothing prevents someone simply calling:
> 
> ```rust
> let unsafe_params = params.raw();
> ```

Yup 🙃 The problem is that FFI is inherently unsafe, and ultimately relies on some degree of convention and manual checking to enforce safety. We could setup `WorkerSdkConversion`/`RawBinding`, but even that's still just a convention, and there's nothing stopping us from just writing a regular `to_worker_sdk` method without going through the traits.

Reply:
Yeah, I agree with @randomPoison that it's probably not worth trying to guard against more specific scenarios, as this is really just a small bit of internal interop code.

Reply:
@jamiebrynes7 Were you satisfied with that explanation? Anything else you wanted to go over? If you're good with it, I can get this merged.

Reply:
Yeah all good with me! :smile: